### PR TITLE
BF(TST): Reuse dataset instance and explicitly reload config upon external changes

### LINKED ISSUE #4363: dataset.config fails to detect changed config file (seems to be master specific)
Origin: https://github.com/datalad/datalad-extensions/pull/7#issuecomment-606725942
```
(git)lena:~datalad/datalad-master[master]git
$> python -m nose -s -v datalad/metadata/tests/test_base.py:test_get_metadata_type
datalad.metadata.tests.test_base.test_get_metadata_type ... HERE0
HERE1:  /home/yoh/.tmp/datalad_temp_test_get_metadata_typet6b5ptfh
FAIL
Versions: appdirs=1.4.3 boto=2.49.0 cmd:7z=16.02 cmd:annex=7.20190819+git2-g908476a9b-1~ndall+1 cmd:bundled-git=2.20.1 cmd:git=2.20.1 cmd:system-git=2.25.1 cmd:system-ssh=8.1p1 exifread=2.1.2 git=3.0.7 gitdb=2.0.6 humanize=0.5.1 iso8601=0.1.12 keyring=18.0.1 keyrings.alt=3.2.0 msgpack=0.5.6 mutagen=1.40.0 requests=2.22.0 tqdm=4.30.0 wrapt=1.11.2
Obscure filename: str=b' "\';a&b&c\xce\x94\xd0\x99\xd7\xa7\xd9\x85\xe0\xb9\x97\xe3\x81\x82 `| ' repr=' "\';a&b&cΔЙקم๗あ `| '
Encodings: default='utf-8' filesystem='utf-8' locale.prefered='UTF-8'
Environment: PATH='/home/yoh/proj/datalad/datalad-master/venvs/dev3/bin:/home/yoh/gocode/bin:/home/yoh/gocode/bin:/home/yoh/bin:/home/yoh/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/sbin:/usr/sbin:/usr/local/sbin' LANG='en_US.UTF-8' GIT_PAGER='less --no-init --quit-if-one-screen' GIT_PYTHON_GIT_EXECUTABLE='/usr/lib/git-annex.linux/git'

======================================================================
FAIL: datalad.metadata.tests.test_base.test_get_metadata_type
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/yoh/proj/datalad/datalad-master/datalad/tests/utils.py", line 691, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/yoh/proj/datalad/datalad-master/datalad/metadata/tests/test_base.py", line 92, in test_get_metadata_type
    assert_equal(get_metadata_type(ds), 'mamboschwambo')
AssertionError: [] != 'mamboschwambo'
```

Happens in `test_get_metadata_type` where  we append explicitly to `.datalad/config` file,

It actually seems to be not git/git-annex version specific -- happens with our stock neurodebian git annex 7.20190819+git2-g908476a9b-1~ndall+1, not sure how was not spotted...

somehow does not happen on maint - there config seems to be nicely reread... I have not spotted any relevant diff in `datalad/metadata/tests/test_base.py` or `datalad/config.py`...

May be some recent changes give ideas @kyleam @mih ? before we try to bisect or figure it out otherwise

### PR DESCRIPTION
**Pull request by yarikoptic**
Fixes #4363 
Alternative to #4364

### CODE REVIEW

File: datalad/metadata/tests/test_base.py [hunk 1]
```diff
@@ -78,15 +78,20 @@
 
 @with_tempfile(mkdir=True)
 def test_get_metadata_type(path):
-    Dataset(path).create()
+    ds = Dataset(path).create()
     # nothing set, nothing found
-    assert_equal(get_metadata_type(Dataset(path)), [])
+    assert_equal(get_metadata_type(ds), [])
     # got section, but no setting
     open(opj(path, '.datalad', 'config'), 'w').write('[datalad "metadata"]\n')
```
Comment (kyleam) [inline datalad/metadata/tests/test_base.py:85]: Just a side note, not related to the fix in this PR: It seems add to me that we use `open(..., 'w')` here.  That clobbers the dataset ID config.

File: datalad/metadata/tests/test_base.py [hunk 2]
```diff
@@ -78,15 +78,20 @@
...
-    assert_equal(get_metadata_type(Dataset(path)), [])
+    # not relying on automagical pick up of changes done
+    # by external powers to the config
+    # see https://github.com/datalad/datalad/issues/4363 for more info
+    ds.config.reload()
+    assert_equal(get_metadata_type(ds), [])
     # minimal setting
     open(opj(path, '.datalad', 'config'), 'w+').write('[datalad "metadata"]\nnativetype = mamboschwambo\n')
```
Comment (kyleam) [inline datalad/metadata/tests/test_base.py:92]: ... and I also don't understand why we would use `w+` here.

File: datalad/metadata/tests/test_base.py (hunk 1 above)
Comment (yarikoptic) [reply datalad/metadata/tests/test_base.py:85]: indeed!

File: datalad/metadata/tests/test_base.py (hunk 2 above)
Comment (yarikoptic) [reply datalad/metadata/tests/test_base.py:92]: so we have both I guess

Comment (kyleam) [reply datalad/metadata/tests/test_base.py:92]: > so we have both I guess

both what?

**Comment by mih** [comment]
Ah, perfect. I was expressing my dislike of #4364 before taking a look here. Thx! I concur with @kyleam that this is preferred!

**Comment by yarikoptic** [comment]
ok, let's merge this one to bring `master` into a good shape in regards to tests.

**Comment by kyleam** [comment]
> ok, let's merge this one to bring `master` into a good shape in regards to tests.

Just for clarity: No one has yet pointed to a datalad core Travis build for master that shows this failure, right?  My understanding is that you first saw it when working on datalad/datalad-extensions#7, and then in gh-4363 you and I were both able to trigger the failure locally on master, though the failure has some flakiness and we didn't yet have luck blaming a specific commit.

File: datalad/metadata/tests/test_base.py (hunk 2 above)
Comment (yarikoptic) [reply datalad/metadata/tests/test_base.py:92]: both config records

Comment (kyleam) [reply datalad/metadata/tests/test_base.py:92]: Are you confusing "w+" with "a"?

Comment (yarikoptic) [reply datalad/metadata/tests/test_base.py:92]: I totally did ;)

**Pull request merged**